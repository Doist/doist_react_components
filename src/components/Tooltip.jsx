import './styles/tooltip.less'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import Popover from './Popover'

class Tooltip extends React.Component {
    constructor(props, context) {
        super(props, context)
        this.state = { visible: false }
    }

    shouldComponentUpdate(nextProps, nextState) {
        // only update on state or prop changes
        return (
            this.state.visible !== nextState.visible ||
            this.props.position !== nextProps.position ||
            this.props.text !== nextProps.text ||
            this.props.hideOnScroll !== nextProps.hideOnScroll ||
            this.props.delayShow !== nextProps.delayShow ||
            this.props.delayHide !== nextProps.delayHide ||
            this.props.gapSize !== nextProps.gapSize ||
            this.props.children !== nextProps.children
        )
    }

    componentWillUnmount() {
        this._clearDelayTimeout()
        this._removeScrollListener()
    }

    _initScrollListener() {
        document.addEventListener('scroll', this._hide, true)
    }

    _removeScrollListener() {
        document.removeEventListener('scroll', this._hide, true)
    }

    _clearDelayTimeout() {
        if (this.delayTimeout) {
            clearTimeout(this.delayTimeout)
        }
    }

    _show = () => {
        this._delayAction(() => {
            this.setState(() => ({ visible: true }))
            if (this.props.hideOnScroll) {
                this._initScrollListener()
            }
        }, this.props.delayShow)
    }

    _hide = () => {
        this._delayAction(() => {
            this._clearDelayTimeout()
            this.setState(() => ({ visible: false }))
            if (this.props.hideOnScroll) {
                this._removeScrollListener()
            }
        }, this.props.delayHide)
    }

    _delayAction(actionFn, delay) {
        this._clearDelayTimeout()
        this.delayTimeout = setTimeout(actionFn, delay)
    }

    render() {
        const {
            allowVaguePositioning,
            wrapperClassName,
            tooltipClassName,
            text,
            children,
            gapSize,
            inverted
        } = this.props

        const wrapperClass = classNames(
            'reactist tooltip__wrapper',
            wrapperClassName
        )
        const tooltipClass = classNames('reactist tooltip__text', tooltipClassName, { inverted })

        if (!text) {
            return <div className={wrapperClass}>{children}</div>
        }

        return (
            <Popover
                visible={this.state.visible}
                trigger={children}
                content={text}
                popoverClassName={tooltipClass}
                wrapperClassName={wrapperClass}
                onMouseEnter={this._show}
                onMouseLeave={this._hide}
                allowVaguePositioning={allowVaguePositioning}
                gapSize={gapSize}
                withArrow
            />
        )

        // return (
        //     <span
        //         className={wrapperClass}
        //         onMouseEnter={this._show}
        //         onMouseLeave={this._hide}
        //         ref={wrapper => (this.wrapper = wrapper)}
        //     >
        //         {this.props.children}
        //         <span
        //             className={tooltipClass}
        //             ref={tooltip => (this.tooltip = tooltip)}
        //         >
        //             <span className="tooltip__text">{this.props.text}</span>
        //         </span>
        //     </span>
        // )
    }
}
Tooltip.displayName = 'Tooltip'
Tooltip.defaultProps = {
    position: 'auto',
    hideOnScroll: true,
    delayShow: 500,
    delayHide: 0,
    allowVaguePositioning: false,
    inverted: false,
    gapSize: 5 // default size of the arrow (see `tooltip.less`)
}
Tooltip.propTypes = {
    /**
     * Position of the tooltip. Defaults to `auto`.
     * `auto` tries to position the tooltip to the top,
     * if there's not enough space it tries to position the tooltip clockwise (right, bottom, left).
     * Setting a distinct value like `right` will always position the tooltip right, regardless of available space.
     */
    position: PropTypes.oneOf(['auto', 'top', 'right', 'bottom', 'left']),
    /**
     * Whether vague positioning is allowed. When set to true the tooltip prefers to be fully visible over being correctly centered.
     */
    allowVaguePositioning: PropTypes.bool,
    /** Text that is displayed inside the tooltip */
    text: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]).isRequired,
    /** Set whether scrolling should hide the tooltip or not. */
    hideOnScroll: PropTypes.bool,
    /** How long to wait after hovering before the tooltip is shown (in ms). */
    delayShow: PropTypes.number,
    /** How long to wait after unhovering before the tooltip is hidden (in ms). */
    delayHide: PropTypes.number,
    /** Children that are wrapped by the toolip. */
    children: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]),
    /** Additional css class that is applied to the wrapper element. */
    wrapperClassName: PropTypes.string,
    /** Additional css class that is applied to the tooltip element. */
    tooltipClassName: PropTypes.string,
    /** Inverted tooltips have a light background with dark text. */
    inverted: PropTypes.bool,
    /** Gap between the tooltip wrapper and the arrow  */
    gapSize: PropTypes.number
}

export default Tooltip
