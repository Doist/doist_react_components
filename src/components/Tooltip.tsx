import React from 'react'
import classNames from 'classnames'

import {
    useTooltipState,
    TooltipInitialState,
    Tooltip as ReakitTooltip,
    TooltipReference,
    TooltipProps as ReakitTooltipProps,
} from 'reakit/Tooltip'
import { PopoverState } from 'reakit/Popover'

import './styles/tooltip.less'

type TooltipProps = Omit<ReakitTooltipProps, 'children'> & {
    children: React.ReactNode
    content: React.ReactNode
    position?: PopoverState['placement']
    gapSize?: number
}

const SHOW_DELAY = 100
const HIDE_DELAY = 100

function useDelayedTooltipState(initialState: TooltipInitialState) {
    const tooltipState = useTooltipState(initialState)
    const delay = useDelay()
    return React.useMemo(
        () => ({
            ...tooltipState,
            show: delay(() => tooltipState.show(), SHOW_DELAY),
            hide: delay(() => tooltipState.hide(), HIDE_DELAY),
        }),
        [delay, tooltipState],
    )
}

function Tooltip({
    children,
    content,
    position = 'top',
    gapSize = 3,
    className,
    ...props
}: TooltipProps) {
    const tooltip = useDelayedTooltipState({ placement: position, gutter: gapSize })

    const child = React.Children.only<React.ReactElement>(children as React.ReactElement)
    if (!content) {
        return child
    }

    return (
        <>
            {/* eslint-disable-next-line @typescript-eslint/no-explicit-any */}
            <TooltipReference {...tooltip} ref={(child as any).ref} {...child.props}>
                {(referenceProps) => React.cloneElement(child, referenceProps)}
            </TooltipReference>
            <ReakitTooltip
                {...tooltip}
                {...props}
                className={classNames('reactist_tooltip', className)}
            >
                {content}
            </ReakitTooltip>
        </>
    )
}

export { Tooltip, TooltipProps }

//
// Internal helpers
//

/**
 * Returns a function offering the same interface as setTimeout, but cleans up on unmount.
 *
 * The timeout state is shared, and only one delayed function can be active at any given time. If
 * a new delayed function is called while another one was waiting for its time to run, that older
 * invocation is cleared and it never runs.
 *
 * This is suitable for our use case here, but probably not the most intuitive thing in general.
 * That's why this is not made a shared util or something like it.
 */
function useDelay() {
    const timeoutRef = React.useRef<ReturnType<typeof setTimeout>>()

    const clearTimeouts = React.useCallback(function clearTimeouts() {
        if (timeoutRef.current != null) {
            clearTimeout(timeoutRef.current)
        }
    }, [])

    // Runs clearTimeouts when the component is unmounted
    React.useEffect(() => clearTimeouts, [clearTimeouts])

    return React.useCallback(
        function delay(fn: () => void, delay: number) {
            return () => {
                clearTimeouts()
                timeoutRef.current = setTimeout(fn, delay)
            }
        },
        [clearTimeouts],
    )
}
