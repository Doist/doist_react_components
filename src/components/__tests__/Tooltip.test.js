import React from 'react'
import { shallow, mount } from 'enzyme'

import Tooltip from '../Tooltip'

describe('Tooltip', () => {
    // Setup ==================================================================
    beforeEach(() => {
        jest.useFakeTimers()
    })
    afterEach(() => {
        jest.useRealTimers()
    })

    it('renders without crashing', () => {
        const tooltip = mount(getTooltip())
        expect(tooltip).toMatchSnapshot()
    })

    it('renders only children if text prop is not defined', () => {
        // eslint-disable-next-line
        console.error = jest.fn() // silence errors caused by unsupplied required prop
        const tooltip = shallow(getTooltip({ text: null }))
        expect(tooltip).toMatchSnapshot()
    })

    it('gets visible on hovering and hides on unhovering', () => {
        const tooltip = shallow(getTooltip())

        tooltip.simulate('mouseEnter')
        jest.runAllTimers()
        expect(tooltip.state().visible).toBe(true)

        tooltip.simulate('mouseLeave')
        jest.runAllTimers()
        expect(tooltip.state().visible).toBe(false)
    })

    describe('Arrow Placement', () => {
        it('draws a bottom arrow when rendering to the top', () => {
            const tooltip = shallow(getTooltip({ position: 'top' }))
            tooltip.simulate('mouseEnter')
            jest.runAllTimers()

            expect(tooltip.find('.arrow_bottom')).toBeTruthy()
        })

        it('draws a left arrow when rendering to the right', () => {
            const tooltip = shallow(getTooltip({ position: 'right' }))
            tooltip.simulate('mouseEnter')
            jest.runAllTimers()

            expect(tooltip.find('.arrow_left')).toBeTruthy()
        })

        it('draws a top arrow when rendering to the bottom', () => {
            const tooltip = shallow(getTooltip({ position: 'bottom' }))
            tooltip.simulate('mouseEnter')
            jest.runAllTimers()

            expect(tooltip.find('.arrow_top')).toBeTruthy()
        })

        it('draws a right arrow when rendering to the left', () => {
            const tooltip = shallow(getTooltip({ position: 'left' }))
            tooltip.simulate('mouseEnter')
            jest.runAllTimers()

            expect(tooltip.find('.arrow_right')).toBeTruthy()
        })
    })

    describe('Scroll Listener', () => {
        it('adds a capturing scrolllistener when becoming visible and when hideOnScroll is true', () => {
            document.addEventListener = jest.fn()
            const tooltip = shallow(getTooltip({ hideOnScroll: true }))

            tooltip.simulate('mouseEnter')
            jest.runAllTimers()
            const callArgs = document.addEventListener.mock.calls[0]
            expect(callArgs[0]).toBe('scroll')
            expect(callArgs[2]).toBe(true)
        })

        it('does not add a scrolllistener when becoming visible and when hideOnScroll is false', () => {
            document.addEventListener = jest.fn()
            const tooltip = shallow(getTooltip({ hideOnScroll: false }))

            tooltip.simulate('mouseEnter')
            jest.runAllTimers()
            expect(document.addEventListener).not.toHaveBeenCalled()
        })

        it('removes a capturing scrolllistener when hiding and when hideOnScroll is true', () => {
            document.removeEventListener = jest.fn()
            const tooltip = shallow(getTooltip({ hideOnScroll: true }))

            tooltip.simulate('mouseLeave')
            jest.runAllTimers()
            const callArgs = document.removeEventListener.mock.calls[0]
            expect(callArgs[0]).toBe('scroll')
            expect(callArgs[2]).toBe(true)
        })

        it('does not remove a scrolllistener when hiding and when hideOnScroll is false', () => {
            document.removeEventListener = jest.fn()
            const tooltip = shallow(getTooltip({ hideOnScroll: false }))

            tooltip.simulate('mouseLeave')
            jest.runAllTimers()
            expect(document.removeEventListener).not.toHaveBeenCalled()
        })

        it('removes scroll listener on unmount', () => {
            document.removeEventListener = jest.fn()
            const tooltip = shallow(getTooltip()).instance()

            tooltip.componentWillUnmount()
            expect(document.removeEventListener).toHaveBeenCalled()
        })
    })

    describe('Component Updates', () => {
        it('updates when the state changes', () => {
            const tooltip = shallow(getTooltip()).instance()

            const currentProps = tooltip.props
            const shouldUpdate = tooltip.shouldComponentUpdate(currentProps, {
                visible: true
            })
            expect(shouldUpdate).toBe(true)
        })

        it('updates when any property changes', () => {
            const tooltip = shallow(getTooltip()).instance()
            const currentState = tooltip.state
            const currentProps = tooltip.props

            const shouldUpdateAfterPositionChange = tooltip.shouldComponentUpdate(
                { ...currentProps, position: 'bottom' },
                currentState
            )
            expect(shouldUpdateAfterPositionChange).toBe(true)

            const shouldUpdateAfterTextChange = tooltip.shouldComponentUpdate(
                { ...currentProps, text: 'new tip' },
                currentState
            )
            expect(shouldUpdateAfterTextChange).toBe(true)

            const shouldUpdateAfterHideOnScrollChange = tooltip.shouldComponentUpdate(
                { ...currentProps, hideOnScroll: false },
                currentState
            )
            expect(shouldUpdateAfterHideOnScrollChange).toBe(true)

            const shouldUpdateAfterDelayShowChange = tooltip.shouldComponentUpdate(
                { ...currentProps, delayShow: 2342 },
                currentState
            )
            expect(shouldUpdateAfterDelayShowChange).toBe(true)

            const shouldUpdateAfterDelayHideChange = tooltip.shouldComponentUpdate(
                { ...currentProps, delayHide: 2342 },
                currentState
            )
            expect(shouldUpdateAfterDelayHideChange).toBe(true)

            const shouldUpdateAfterGapSizeChange = tooltip.shouldComponentUpdate(
                { ...currentProps, gapSize: 10 },
                currentState
            )
            expect(shouldUpdateAfterGapSizeChange).toBe(true)

            const shouldUpdateAfterChildrenChange = tooltip.shouldComponentUpdate(
                { ...currentProps, children: <span>New Content</span> },
                currentState
            )
            expect(shouldUpdateAfterChildrenChange).toBe(true)
        })

        it('does not update when state and properties did not change', () => {
            const tooltip = shallow(getTooltip()).instance()

            const currentState = tooltip.state
            const currentProps = tooltip.props
            const shouldUpdate = tooltip.shouldComponentUpdate(
                currentProps,
                currentState
            )
            expect(shouldUpdate).toBe(false)
        })
    })

    it('sets ref of tooltip and wrapper after mounting', () => {
        const tooltip = mount(getTooltip()).instance()

        expect(tooltip.wrapper).toBeTruthy()
        expect(tooltip.tooltip).toBeTruthy()
    })

    // Helpers ================================================================
    const getTooltip = (props = {}) => (
        <Tooltip text="tip" {...props}>
            <span>Wrapped Content</span>
        </Tooltip>
    )
})
