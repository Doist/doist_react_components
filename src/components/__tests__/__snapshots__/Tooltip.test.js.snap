// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders without crashing 1`] = `
<div
  className="reactist tooltip__wrapper"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <span>
    Wrapped Content
  </span>
  <div
    className="reactist tooltip"
  >
    tip
  </div>
</div>
`;
